### DISCUSSION TITLE
How to get started using tui

### DISCUSSION THREAD

**Discussion by winitzki** [oyvindberg/tui-scala#52] [General]
Hello,

I'm intrigued by text-based UI but I couldn't find any documentation about tui-rs or tui-scala. How can I get started? Say, I want to write a Scala app that sends commands to a Web server using a text-based UI. As input I need drop-down choice boxes, tick boxes, and free input fields, say.

I'm also confused by the statement that "this library does not implement any events". How can a UI library work without implementing events? Is there a runloop, is there a declarative description of the logic, or how does it work?

**Comment by oyvindberg** [comment]
Hi there

See any of the examples in the readme file to get started, plus you need to add the dependency to you build. I'll make the readme file a bit clearer.

There is no dropdown widget and so on here yet unfortunately. I'd love to have it though.

It does have events, but maybe not at the level you probably would expect them. You write a top-level loop which handles it

**Comment by winitzki** [comment]
Thank you!

Let me just add two comments. It would help if the library were published to sonatype, then at least it would be very easy to check out the examples and start playing with it.

Second, I'm looking at [Elm](https://elm-lang.org/) and I think the Elm architecture will fit here. It's a very straightforward thing where an interactive program is a pure value of type `(M, M => V[E], M => E => M)` where M is a model, E is an event type, and V[E] is a view whose elements are parameterized by the event type. Then there is a runtime engine that takes this value and runs it. But the user code never has any side effects.

The view can use Cassowary constraints, widgets, etc., and the Elm architecture makes it declarative; there are no threads or race conditions to think about.

I would be interested in about 1 year to create a project implementing Elm architecture for text-based UIs.

**Reply by oyvindberg** [reply]
It is published to sonatype! https://mvnrepository.com/artifact/com.olvind.tui/tui

Re: elm, that's an interesting discussion. I have made some attempts in the direction of making the widgets into something reminiscent of react components, which is a step in the direction to the elm architecture as I see it. 

The main problem I have so far is that some parts of rendering to terminal is inherently a bit more imperative and cell-oriented than is rendering to a DOM. In the definition of widgets you necessarily need to know something about coordinates, and render things with some manual layouting. 

That observation led me towards investigating a separation of high/low-level "components", where when you create new widgets (like a checkbox) you do it imperatively, whereas when you use those low-level components you do it at a high level, where you don't see or care about coordinates and so on. basically you can just compose them just as you do react components.

See some early work and discussion in #50 .

One problem that I did come across when working towards "components" (whatever that ends up being in tui-scala), was that of events. I think the elm model could be a great match for modelling that here.

Tell me if you're interested in making a POC for this or something in that direction, I'd be willing to help and discuss. Other than that, continued work on tui-scala is some months into the future, as I have my focus on another project I haven't published yet.

**Reply by winitzki** [reply]
Unfortunately I can't really follow the discussion in #50 and elsewhere. I mostly don't understand what they are talking about, it's too involved in the technical details that I'm not familiar with.

You say that rendering to terminal is more imperative than rendering to DOM. This is true but the Elm architecture frees you from all imperativity in the user code. All imperative code, all asynchronous code, all multithreaded code is moved into the Elm runtime and is completely invisible to the user. This is a great achievement of the Elm language. The UI programmer no longer has any concerns about threads, the state of the screen, the cursor position, the order of updates and concurrency, and so on. The UI programmer only describes what layouts correspond to various values of the Model type (`M`), what events are possible given a layout, and what model updates correspond to each possible event.

"React" is not a faithful implementation of the Elm architecture; rather, the Elm architecture would be there with React + Flux + Redux and with no components. Actually, the plain Elm architecture is not modular: the entire UI of your application is a single big "component". The UI of course is made of parts but those parts are only present as parts of the View type. There is a single model for the entire application, and a single view, and a single type for all events.

Thank you for publishing the tui library,  I will try to do something with that artifact and let you know in case nothing works.

My implementation plan for a POC would be like this:

- Define a couple of View types. Label and Button, say.

```scala
sealed trait View[E]
final case class Label[E](text: String) extends View[E]
final case class Button[E](text: String, event: E) extends View[E]
```

- Define simple layout combinators (tile vertical, tile horizontal) to combine several View types together into a larger View.
```scala
final case class TileH[E](left: View[E], right: View[E])  extends View[E]
final case class TileV[E](top: View[E], bottom: View[E])  extends View[E]
```

- Define the type of "simple Elm program": 

```scala
type SimpleElmProgram[M, V[_], E] = (M, M => V[E], M => E => M)
```
 - Implement an "Elm runner", which starts a runloop.

Given an Elm program, we run one step of that program, which contains a function `M => V[E]`. We substitute the current value of the model `M` and get the resulting view V[E]. We then draw the corresponding picture using `tui` on the screen, creating a given layout with labels and buttons. Then there will be some kind of waiting for keypresses or mouse clicks. We must determine which part of the view was clicked on, or in which part of the view the keypress went. The value of type `E` will be extracted from that part of the view. Then we run the updater function `M => E => M` to get a new value of the model. Then we repeat. This is the runloop.

I don't foresee any problems implementing this scheme.

Sample Elm program: We have a label and two buttons: "Increment" and "Reset". The label shows how many times we clicked "Increment", initially the label shows "0 clicks". Click "Reset" and it again resets the counter to 0.

```scala
type M = Int // Counter.
type E = Boolean // Two buttons. "true" means "Increment", "false" means "Reset".
val display: M => View[E] = m => TileV(Label(s"$m clicks"), TileH(Button("Increment", true), Button("Reset", false)))
val update: M => E => M = m => e => e match { // Compute new model value given an event `e`.
  case true => m + 1
  case false => 0
}
val program: SimpleElmProgram[M, View, E] = (0, display, update)
runElm(program) // Start the runloop.
```

The user's code is a pure value of type `SimpleElmProgram`. All the imperative code, all the complexity of event handling is encapsulated by the runtime engine (`runElm`).

**Comment by oyvindberg** [comment]
I think this is interesting. 

We need to find a way to componentize the widgets, including callbacks. So a user can easily compose components without getting lost in individual coordinates and complicated state management. It needs to be structured enough that we can get tab navigation for instance "for free". This is the main thing I want to accomplish in the short term. Note that some amount of imperative code is OK in my book to fulfill these requirements, especially if it makes it easier to use the library for beginners.

I'm sure you can discuss details of elm and react architecture for a while, but for my purposes here let's just talk loosely about a components as pure rendering function + state management (including callbacks). 

The elm architecture is definitely a possible direction, though the scope of implementing it look like a full rewrite of the rendering and widgets layers. Maybe any componentization of the library would end up there actually.

I wanted to drop a few links as possible inspiration apart from elm proper, I'm sure there are tons more.
- [scalm](github.com/julienrf/scalm/) is a scala-elm thing, maybe we could save some time by rebasing that on top of tui-scala foundations.
- in the [roadmap](https://github.com/oyvindberg/tui-scala/issues/15) I point to `cursive` as a better design, in that is handles callbacks and so on. I think we should have a look at that as well
- [interim](https://github.com/JD557/interim), which is not terminal-based, but looks like it could be with a few changes.


### challenges
- there is some mutability in the terminal, exactly as there is in the DOM.
- crossterm and tui-scala (ab-)uses the mutability to patch terminals cells. for instance in `main` you may paint all cells in the coordinate system as having a given background color, and then render widgets on top which will render and inherit that background color. I worked towards removing this in #26 , but closed it in the end as I thought it was a huge step backwards in terms of usability. But we need something like this.
- All the current widgets are written in terms of (relative) coordinates. Working towards abstracting this away will be complicated and work-intensive. I think we can look towards `cursive` for how it does layouting in this case, where a given component exposes minimum amount of space it needs.
- callbacks maybe. It needs to be implemented from scratch. In order to implement it we need to keep some state about the currently rendered component hierarchy, so we can for instance go from terminal coordinate to the component which rendered it to handle mouse clicks.

### where to go from here

If you're interested in working on this I'd be very happy to see a proof of concept, where we for instance create an elm-like architecture, some widget like a checkbox which is controllable by both mouse and keyboard, and a user program which renders a bunch of checkboxes to test layouting and user interaction. 

I hope that such a prototype can be easy on type parameters and "advanced" features. Ideally at a level where it could be ported to java, although not a hard requirement. It's a big plus if it works both for scala 2 and 3.

**Comment by winitzki** [comment]
First off, I did not mean to say that Elm is the only way to go. I see Elm as a high-level interface that can be slapped on top of any UI library that has callbacks, mutable DOM, or whatever.

First, a UI library must somehow, most likely imperatively, draw the views, manage layouts, and catch user events. Then, independently of that, one needs to implement a runloop and some higher-level "widgets". The Elm architecture is one possibility among many. The UI library should be Elm-agnostic and should concentrate on good maintainability of the low-level code.

An Elm runner will be implemented entirely through the public API of a UI library. So, Elm-related considerations are not needed while you are still working out the low-level details of the UI library: how to implement overlapping widgets, how to minimize redrawing, how to detect which widget got clicked or has focus, and so on.  The UI library needs to be able to draw a given layout with widgets, handle focus, and catch events from widgets. To do that, the UI library may use any mechanism - callbacks, mutable global variables, virtual DOM or just a big screen buffer, whatever works.  The UI library may implement its own runloop but it is not necessary at that level. I leave it entirely to your judgment (certainly you have more experience than I) to decide what to do with tui-scala in terms of low-level implementation.

I am not familiar with the "immediate mode GUI" and I will need to read up on that. 

I wanted to make a proof of concept with the code I showed above in my previous comment. I created a small project, imported tui-scala and pasted my code and some example code from the tui-scala repo. I hoped to figure out how to draw the widgets and how to catch events from them. But I couldn't figure out anything. The code of the widgets is too imperative and it is not clear how to use tui-scala, and there is no documentation or tutorial or explanation about what "widgets" do, what is going on when you call `f.renderWidget`, `f.renderStatefulWidget`, etc. After looking at the examples in the repo, I'm still not able to guess how to draw one label and two buttons in a simple layout, and how to catch the "button clicked" events (i.e. not just a keypress with a fixed key).

So, getting anything done will require a few continuous weeks of going through your entire code, until I understand how it works. But I have perhaps one half-hour per week to look at this project. That is why this discussion thread has the title "how to get started using tui".

**Comment by winitzki** [comment]
Here is an example from the `InterIm` library that you pointed me to. This code shows a generic kind of runloop. It is assumed that `application` manages its internal state imperatively.

https://github.com/JD557/interim/blob/master/examples/release/intro.md

```scala
def application(inputState: InputState) = ??? // Our application code

while(true) {
  val input: InputState = backend.grabInput() // Grab input from the backend
  val (renderOps, _) = application(input) // Generate render ops
  backend.render(renderOps) // Send render operations to the backend
}
```

It is clear from this example how to implement any high-level GUI architecture, such as Elm or "immediate UI" or anything else, using just the public API of a "backend". The "backend" just needs to be able to grab user events and to render a given view in some way.

**Comment by winitzki** [comment]
I just found `jexer`, another text-mode UI framework.

https://gitlab.com/AutumnMeowMeow/jexer/-/wikis/high-level-design
